module/zfs/metaslab.c

```diff
@@ -236,12 +265,81 @@ metaslab_class_destroy(metaslab_class_t *mc)
 	ASSERT(mc->mc_deferred == 0);
 	ASSERT(mc->mc_space == 0);
 	ASSERT(mc->mc_dspace == 0);
+#ifdef ZFS_DEBUG
+	for (int t = 0; t < TXG_SIZE; t++)
+		ASSERT0(mc->mc_alloc_stats_dirty[t]);
+#endif
 
+	avl_destroy(&mc->mc_alloc_tree);
 	refcount_destroy(&mc->mc_alloc_slots);
 	mutex_destroy(&mc->mc_lock);
+	mutex_destroy(&mc->mc_alloc_lock);
 	kmem_free(mc, sizeof (metaslab_class_t));
 }
 
+void
+metaslab_class_enable_stats(metaslab_class_t *mc, int64_t class_stats_obj)
+{
+	spa_t *spa = mc->mc_spa;
+	objset_t *mos = spa_meta_objset(spa);
+	dmu_buf_t *dbp;
+	int error;
+
+	ASSERT(class_stats_obj != 0);
+	ASSERT0(mc->mc_alloc_stats_obj);
+
+	error = dmu_bonus_hold(mos, class_stats_obj, FTAG, &dbp);
+	if (error) {
+		cmn_err(CE_WARN, "can't hold mc_alloc_stats_dbuf");
+		return;
```

Comment:
can we really survive this error?  The obj exists, but if we don't load it we won't be able to update it, right?  Then the stats will be wrong.

Reply:
I can't think of a reason why the hold would ever fail. A `zdb -b` audit would reveal the stats were wrong (and fail  a ztest run).